src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -883,35 +895,60 @@ export class BundleDataClient {
-                // so this fill is validated.
+                // so this fill can no longer be filled on-chain.
                 v3RelayHashes[relayDataHash].fill = fill;
                 if (fill.blockNumber >= destinationChainBlockRange[0]) {
                   const fillToRefund = await verifyFillRepayment(
                     fill,
                     destinationClient.spokePool.provider,
-                    v3RelayHashes[relayDataHash].deposit!,
+                    v3RelayHashes[relayDataHash].deposits![0],
                     allChainIds
                   );
                   if (!isDefined(fillToRefund)) {
+                    // We won't repay the fill but the depositor has received funds so we don't need to make a
+                    // payment.
                     bundleUnrepayableFillsV3.push(fill);
                     // We don't return here yet because we still need to mark unexecutable slow fill leaves
                     // or duplicate deposits. However, we won't issue a fast fill refund.
                   } else {
                     v3RelayHashes[relayDataHash].fill = fillToRefund;
                     validatedBundleV3Fills.push({
                       ...fillToRefund,
-                      quoteTimestamp: v3RelayHashes[relayDataHash].deposit!.quoteTimestamp, // ! due to assert above
+                      quoteTimestamp: v3RelayHashes[relayDataHash].deposits![0].quoteTimestamp, // ! due to assert above
+                    });
+
+                    // Now that we know this deposit has been filled on-chain, identify any duplicate deposits
+                    // sent for this fill and refund them to the filler, because this value would not be paid out
+                    // otherwise. These deposits can no longer expire and get refunded as an expired deposit,
+                    // and they won't trigger a pre-fill refund because the fill is in this bundle.
+                    // Pre-fill refunds only happen when deposits are sent in this bundle and the
+                    // fill is from a prior bundle. Paying out the filler keeps the behavior consistent for how
+                    // we deal with duplicate deposits regardless if the deposit is matched with a pre-fill or
+                    // a current bundle fill. If the fill is a slow fill,
+                    const duplicateDeposits = v3RelayHashes[relayDataHash].deposits!.slice(1);
+                    duplicateDeposits.forEach((duplicateDeposit) => {
+                      // If fill is a slow fill, refund deposit to depositor, otherwise refund to filler.
+                      if (isSlowFill(fill)) {
+                        updateExpiredDepositsV3(expiredDepositsToRefundV3, duplicateDeposit);
+                      } else {
+                        validatedBundleV3Fills.push({
+                          ...fillToRefund,
+                          quoteTimestamp: duplicateDeposit.quoteTimestamp,
```

Comment:
Something I thought about that might not be a big deal: the LP fee charged on these duplicate fill refunds is based on the duplicate deposit's quote timestamp. Quote timestamps can differ between duplicate deposits because they are not part of the relay data hash. I think this is OK but want to flag it for careful review

Reply:
Yeah, I think this works.

We just need to make sure we are consistent about _which_ deposit is used to determine the slow fill payout amount (I think it would have to be the first).

Reply:
yep. so do you agree then:
- lp fees charged on fill refunds arising from duplicate deposits are based on the duplicate deposit's quote timestamp
- lp fees for slow fills match only the first deposit